### ISSUE TITLE
Empty response for file type, when produces = xml

### ISSUE THREAD

**Issue by weidongxu-microsoft** [Azure/autorest.modelerfour#172]
**Describe the issue**

Empty response for operation with file type response, when `produces = application/xml`.
A nearby operation with `produces = application/octet-stream` appears to be correct.

https://github.com/Azure/azure-rest-api-specs/blob/master/specification/web/resource-manager/Microsoft.Web/stable/2019-08-01/WebApps.json#L7226
```
        "produces": [
          "application/xml"
        ],

          "200": {
            "description": "OK",
            "schema": {
              "type": "file"
            }
          },
```

YAML
```
    responses:
    - language:
        default:
          name: ''
          description: ''
      protocol:
        http:
          statusCodes:
          - '200'
```

IMHO, this seems should still be binary response, i.e. `binary: true`.

** REQUIRED ** 
**ModelerFour** version:4.6.200
**AutoRest-core** version:3.0.6220
**CommandLine** :
OpenAPI/Swagger file:https://github.com/Azure/azure-rest-api-specs/blob/master/specification/web/resource-manager/Microsoft.Web/stable/2019-08-01/WebApps.json

**Comment by fearthecowboy** [comment]
This is an interesting notion.

In this case, we're told "XML"; but the inference here is that the payload is expected to be treated as a binary blob instead of a object deserialization operation.

Yeah, I gotta think about this a smidgen; the inference based on the body media type alone isn't sufficient.

**Comment by fearthecowboy** [comment]
Tangentially, there's already a problem with that spec -- it fails to say produces `application/json`  for the error response. 

@lmazuel @daviwil -- how do you feel about this...

**WARNING: RAMBLING AHEAD.**

- The input clearly specifies that the content type expected is `application/xml` 
- Which I infer as a XML payload (which I'd look to have a schema and deserialize it)
- except there is a `schema: { type: file }` for the response schema.

I _could_ do something to go "_oh, they actually want the output as a blob_" and create the response that way.  

Which introduces an additional problem - if they specified produces `application/json` and `application/xml` , there is no actual connection between the `produces` and the individual status code response definitions in `responses`. 

``` js
"responses": {
  "200": {
    "description": "OK",
    "schema": { 
      "type": "file"
    }
  },
  "default": {
    "description": "App Service error response.",
    "schema": {
      "$ref": "./CommonDefinitions.json#/definitions/DefaultErrorResponse"
    }
  }
}

```
If the `200` response specifies `{type: file}`  I'm supposed to infer `application/xml` **isn't** a deserialization response. I'd have to blindly make all `200` outputs are binary, and construct the responses so that the actual media type for that response is really anything. 

The terrible part is that the OAI2-to-OAI3 up conversion has no idea which produces goes with which response, it basically table-joins them all together, and the modelerfour has to filter out combinations that are not good (exclude json/xml/form media types from binary responses and vice-versa).

If I change it so that the high-order bit is the `schema` (at least for binary), then it means that the response for binary will accept all `produces` media types as valid matches. 

Which *may* or *may not* be the end of the world.  It does unnecessarily add the possibility of accepting unrelated media types for a given binary response.   And would preclude the ability to deliberately use OAI3 and different media types with the same response status code to give actual different outputs (ie, `200+application/json` => deserialize vs `200+image/png` => blob -- which I'd hope we don't see, but who knows )

_Unless_ I internally tag the responses during upconversion with something like `x-ms-upconvert` and when I see that allow relaxed/lazy binary media-type matches. 

urg.

The other thought, would be to maybe add support for something like `"x-ms-produces": [ ... ]` in swagger responses, which would allow us to gain the fidelity that we have in OAI3. We'd only *want* to use that when we have the case where someone wants XML, JSON, or FORM as a binary response.  That would let me make the presence of  `schema: { type: file }` be the high-order bit, and then I wouldn't have to guess when to filter out things from binary types ever, and I would only ever filter illogical binary types out of schema-deserialization responses.  

Ok, I like that. I propose :
- if the response has `schema: { type: file }`,   I **never** filter the media types for that response. 
- add support for `x-ms-produces` at the response-level in swagger so that it is *possible* to deliberately specify a narrow set of produces for a binary (ie, just `200 XML` => binary  instead of allowing all (`200 XML` or `200 JSON` )=> binary ).
- **maybe**: if the response doesn't have a `schema `, but does specify a binary media type, we assume a binary stream anyway (as if they specified `schema: { type: file }`.

**Comment by daviwil** [comment]
Your proposal at the end sounds good to me!  Maybe @lmazuel knows of some edge cases that might exercise it a bit?

**Comment by weidongxu-microsoft** [comment]
I see. It more or less a OAI2 issue that not able to specify different media type for different response.

I think the proposal makes sense. "schema=file" would take priority. Then for other status code without "schema=file" it uses the media-type provided. If that is still not enough (in this case, default probably need JSON), use the `x-ms-produces` to introduce similar feature as OAI3.

BTW, something that complicate like 200 XML=>XML schema while 200 JSON=>binary is probably not even have a chance in current java azure-core impl, which basically only handle 2 type, success to certain type (json or xml or file), and default/error to another (json or xml).

**Comment by lmazuel** [comment]
That's sounds fair to me as well.

Some notes on the Python side, if this brings some thought:
Python generator cares about only one thing: is this a stream or not? I never look at the produces. This creates two kinds of code:
- I expect a **schema**: use `Content-Type` to deserialize the bytes into a JSON tree, or a XML tree or even plain text (if content-type was `text/plain`). I will raise an exception only if the data are not UTF-8 valid
- I expect a **stream**, ignore Content-Type and stream that

I never looked into the produces or media-types, and I don't want to.

I could if I have information that some content negotiation was possible generates something like:
```python
if response.headers["content-type"] is binary:
   return stremable(response)
else:
   return deserialize(response, 'Foo')  # deserialize into model Foo
```

I think your proposal gives me this ability, so I'm fine

**Comment by fearthecowboy** [comment]
Yes, it basically handles the decision making for you, so should be ok.

**Comment by fearthecowboy** [comment]
I've published `4.9.225` -- it's very new, give it a shot. (it rolls up a lot of small changes, so hopefully no big bumps)

**Comment by weidongxu-microsoft** [comment]
Tried it. Works good. Thanks.
